rotkehlchen/exchanges/bitfinex.py

```diff
@@ -0,0 +1,933 @@
+    def _api_query_paginated(  # pylint: disable=no-self-use
+            self,
+            options: Dict[str, Any],
+            case: Literal['asset_movements'],
+            currency_map: Dict[str, str],
+    ) -> List[AssetMovement]:
+        ...
+
+    def _api_query_paginated(
+            self,
+            options: Dict[str, Any],
+            case: Literal['trades', 'asset_movements'],
+            currency_map: Dict[str, str],
+    ) -> Union[List[Trade], List[AssetMovement]]:
+        """Request a Bitfinex API v2 endpoint paginating via an options
+        attribute.
+
+        Pagination attribute criteria per endpoint:
+          - trades: from `start_ts` to `end_ts` (as offset) with limit 2500.
+          - movements: from `start_ts` to `end_ts` (as offset) with limit 1000.
+
+        Error codes documentation:
+        https://docs.bitfinex.com/docs/abbreviations-glossary#errorinfo-codes
+
+        Rates limit documentation:
+        https://docs.bitfinex.com/docs/requirements-and-limitations#rest-rate-limits
+        """
+        deserialization_method: DeserializationMethod
+        endpoint: Literal['trades', 'movements']
+        response_case: Literal['trades', 'asset_movements']
+        if case == 'trades':
+            endpoint = 'trades'
+            response_case = 'trades'
+            deserialization_method = self._deserialize_trade
+            expected_raw_result_length = API_TRADES_MIN_RESULT_LENGTH
+            id_index = 0
+            timestamp_index = 2
+        elif case == 'asset_movements':
+            endpoint = 'movements'
+            response_case = 'asset_movements'
+            deserialization_method = self._deserialize_asset_movement
+            expected_raw_result_length = API_MOVEMENTS_MIN_RESULT_LENGTH
+            id_index = 0
+            timestamp_index = 5
+        else:
+            raise AssertionError(f'Unexpected {self.name} case: {case}')
+
+        call_options = options.copy()
+        limit = options['limit']
+        results: Union[List[Trade], List[AssetMovement]] = []  # type: ignore # bug list nothing
+        last_result_id = None
+        last_result_timestamp = None
+        retries_left = API_REQUEST_RETRY_TIMES
+        while retries_left >= 0:
+            response = self._api_query(
+                endpoint=endpoint,
+                options=call_options,
+            )
+            if response.status_code != HTTPStatus.OK:
+                # NB: check if the rate limits have been hit (response JSON as dict)
+                try:
+                    response_dict = rlk_jsonloads_dict(response.text)
+                except (AssertionError, JSONDecodeError):
+                    msg = (
+                        f'Error decoding {self.name} {case} unsuccessful response JSON as dict. '
+                        f'Continue decoding unsuccessful response as a JSON list.'
+                    )
+                    log.debug(msg, options=call_options)
+                else:
```

Comment:
Not sure if your suggestion would work for all the different cases, let me explain. The response can be:
- A `list` (entry) if status code is 200
- Either a `dict` (rate limit error) or a `list` (any other error) if status code is not 200. That means we must decode each non-successful response as a dict first and as a list.
  - If decoding as a dict fails, re-route it to `_process_unsuccessful_response` (it will decode it as a list).
  - If decoding as a dict works but does not contain the error limit message (this should not happen), we log it (for being aware) and send it to `_process_unsuccessful_response` (it will raise JSONDecodeError).
  - If decoding as a dict works and has the rate limit message, trigger the retry logic.

It should `continue` only if it falls in the retry logic and there are retries left.

Reply:
I could do this, but initializing first `response_dict` as empty dict and not logging the "unexpected" case (a dict without an "error" key). Let me know!
```python
            if response.status_code != HTTPStatus.OK:
                # NB: check if the rate limits have been hit (response JSON as dict)
                response_dict = {}
                try:
                    response_dict = rlk_jsonloads_dict(response.text)
                except (AssertionError, JSONDecodeError):
                    msg = (
                        f'Error decoding {self.name} {case} unsuccessful response JSON as dict. '
                        f'Continue decoding unsuccessful response as a JSON list.'
                    )
                    log.debug(msg, options=call_options)

                if response_dict.get('error', None) == API_RATE_LIMITS_ERROR_MESSAGE:
                    if retries_left == 0:
                        raise RemoteError(
                            f'{self.name} {case} request failed after retrying '
                            f'{API_REQUEST_RETRY_TIMES} times.',
                        )
                    # Trigger retry
                    log.debug(
                        f'{self.name} {case} request reached the rate limits. Backing off',
                        seconds=API_REQUEST_RETRY_AFTER_SECONDS,
                        options=call_options,
                    )
                    retries_left -= 1
                    gevent.sleep(API_REQUEST_RETRY_AFTER_SECONDS)
                    continue

                return self._process_unsuccessful_response(
                    response=response,
                    case=response_case,
                )
```

Reply:
> Either a dict (rate limit error) or a list (any other error) if status code is not 200. That means we must decode each non-successful response as a dict first and as a list if it failed the former.

Okay in that case then the approach is really complex. Please refactor the way you do it and avoid duck typing with exceptions for control handling. An exception should be reserved for an error and in your case you seem to use it also for valid cases.

You can simply use `rlk_jsonloads()` which returns either a list or a dict and then treat the response dependin on what is returned. That would be far more readable.

Reply:
@LefterisJP I've amended it in my last commit:
https://github.com/rotki/rotki/pull/1968/commits/36e53126d482a9dc46d1c9303742ba1349acddba#diff-e8492f0f885eb2b81359e9c50271b2256ec236f1bbc7e7237414e0f0cf3bfe9bR275

I wish it is the last time I use exceptions for control handling :(

Btw, if I create the following variable, all `return []` won't need to be type ignored, but it is very mypy specific.
```python
empty_results: Union[List[Trade], List[AssetMovement]]  = []
```

Reply:
Can you also point to the mypy bug that you are referring to with the empty list? It rings a bell but not sure where it is. I amreorting bugs to both mypy and pylint when I find them. (most probably others have already -- but sometimes I find something new)

Reply:
I could be wrong, but this is what I've experimented/found. There are 2 problems related with `List[<nothing>]`.

One is when the variable initialised as empty list requires an annotation. Mypy moans with "expression has type "List[<nothing>]".

For instance:
https://github.com/python/mypy/issues/2164
https://stackoverflow.com/questions/58906541/incompatible-types-in-assignment-expression-has-type-listnothing-variabl

The other one is when the function return type is "List[type]" and you just do `return []`. As there is no annotation it is `List[<nothing>]` as well. This can be sorted by having a properly annotated variable like the `empty_results` I pointed on my comment above. 

For instance:
https://github.com/python/mypy/issues/6613
https://github.com/python/mypy/issues/7780
https://github.com/python/typing/issues/157
https://github.com/python/mypy/issues/3283

I repeat I could be wrong :S

Reply:
This is something we should look out of this PR